packages/@lwc/engine-core/src/framework/template.ts

```diff
@@ -113,6 +115,59 @@ function validateLightDomTemplate(template: Template, vm: VM) {
     }
 }
 
+enum FragmentCache {
+    HAS_TOKEN = 1, // 2^0
+    HAS_SCOPED_STYLE = 2, // 2^1
+    SHADOW_MODE_SYNTHETIC = 4, // 2^2
+}
+
+export function parseFragment(strings: string[], ...keys: number[]): () => Element {
+    const cache = Object.create(null);
+
+    return function (): Element {
+        const {
+            context: { hasScopedStyles, stylesheetToken },
+            shadowMode,
+        } = getVMBeingRendered()!;
+        const hasStyleToken = !isUndefined(stylesheetToken);
+        const isSyntheticShadow = shadowMode === ShadowMode.Synthetic;
+
+        const cacheKey =
+            (hasStyleToken ? FragmentCache.HAS_TOKEN : 0) | // @todo: can this change?
+            (hasScopedStyles ? FragmentCache.HAS_SCOPED_STYLE : 0) | // if styles are reset, we need to recompute.
+            (isSyntheticShadow ? FragmentCache.SHADOW_MODE_SYNTHETIC : 0); // mixed mode support.
```

Comment:
I'm not sure this is necessary. There are four cases to consider:

| | Native | Synthetic |
| --- | --- | --- | 
| Has styles | no token | token |
| No styles | no token | no token|

This is because of https://github.com/salesforce/lwc/pull/2670, which sets the `stylesheetToken` to `undefined` if there are no styles.

So I think you can just cache based on `FragmentCache.HAS_TOKEN`. I don't think the other ones are necessary.

Reply:
Wait no, I'm wrong. The `stylesheetToken` is passed in even for a native component. It just doesn't get rendered to the DOM. Disregard.